src/shared/cloudformation/templateRegistry.ts

```diff
@@ -0,0 +1,143 @@
+    if (!templateRegistry) {
+        throw new Error(
+            'Template Registry not initialized. Extension code should call activate() from shared/cloudformation/activation, test code should call setTemplateRegistry().'
+        )
+    }
+
+    return templateRegistry
+}
+
+export function setTemplateRegistry(registry: CloudFormationTemplateRegistry | undefined): void {
+    templateRegistry = registry
+}
+
+export interface CloudFormationTemplateRegistry {
+    registeredTemplates: Map<string, CloudFormation.Template>
+    getRegisteredTemplate(templatePath: string): CloudFormation.Template | undefined
+    addTemplateToTemplateData(templatePath: vscode.Uri): Promise<void>
+    removeTemplateFromRegistry(templatePath: vscode.Uri): void
+}
+
+export class DefaultCloudFormationTemplateRegistryListener implements FileWatcherListener {
+    public constructor(private readonly registry: CloudFormationTemplateRegistry) {}
+
+    /**
+     * Function used by a vscode.FileSystemWatcher to handle an onDidChange event.
+     * Adds template to data map. TODO: Emit event
+     * @param templateUri vscode.Uri for the changed template file
+     */
+    public async onListenedChange(templatePath: vscode.Uri): Promise<void> {
+        await this.registry.addTemplateToTemplateData(templatePath)
+        // TODO: fire event
+    }
+
+    /**
+     * Function used by a vscode.FileSystemWatcher to handle an onDidCreate event.
+     * Adds template to data map. TODO: Emit event
+     * @param templateUri vscode.Uri for the created template file
+     */
+    public async onListenedCreate(templatePath: vscode.Uri): Promise<void> {
+        await this.registry.addTemplateToTemplateData(templatePath)
+        // TODO: fire event
+    }
+
+    /**
+     * Function used by a vscode.FileSystemWatcher to handle an onDidDelete event.
+     * Removes template from data map. TODO: Emit event
+     * @param templateUri vscode.Uri for the deleted template file
+     */
+    public async onListenedDelete(templatePath: vscode.Uri): Promise<void> {
+        this.registry.removeTemplateFromRegistry(templatePath)
+        // TODO: fire event
+    }
+
+    /**
+     * Disposes the  vscode.EventEmitters tied to this class
+     * // TODO: Add event emitters to the class and dispose of them here
+     */
+    public dispose(): void {
+        // TODO: Add event emitters to be disposed
+    }
+}
+
+export class DefaultCloudFormationTemplateRegistry implements CloudFormationTemplateRegistry {
+    private readonly templateRegistryData: Map<string, CloudFormation.Template>
+
+    public constructor () {
+        this.templateRegistryData = new Map<string, CloudFormation.Template>()
+    }
+
+    /**
+     * Returns the registry's data in Map form.
+     * Key: template file name, Value: TemplateData object reflecting data in template
+     * TODO: Do we want to return this as a map, an array, or a JS object?
```

Comment:
I recommend not leaking out the implementation (in this case, how it is stored). This makes it harder to change the implementation if we want to, and there is risk of caller mutating the object.

Some options:
* return just a list of templates (no filenames)
* return a List of structures that each contain filename and template